op-service/sources/prefetching_eth_client.go

```diff
@@ -0,0 +1,192 @@
+package sources
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+type PrefetchingEthClient struct {
+	inner            *EthClient
+	prefetchingRange uint64
+}
+
+// NewPrefetchingEthClient creates a new [PrefetchingEthClient] with the given underlying [EthClient]
+// and a prefetching range.
+func NewPrefetchingEthClient(inner *EthClient, prefetchingRange uint64) (*PrefetchingEthClient, error) {
+	return &PrefetchingEthClient{
+		inner:            inner,
+		prefetchingRange: prefetchingRange,
+	}, nil
+}
+
+func (p *PrefetchingEthClient) FetchWindow(ctx context.Context, start, end uint64) {
+	for i := start; i <= end; i++ {
+		// Ignoring the error and result as this is just prefetching
+		// The actual fetching and error handling will be done when the data is requested
+		// FLAG FOR CODE REVIEW: Should we be creating new contexts here? The same context is just passed around everywhere.
```

Comment:
Yes, definitely need new contexts. Contexts should be expected to be closed as soon as the original call returns. They're not safe for background jobs.

What you can do is attach a context, with cancel function, to the PrefetchingClient struct. And then when firing off jobs, use that as a resource context (and you can wrap the context with a timeout per job).
And then add a `Close()` method to the PrefetchingClient, which should close the underlying RPC client, as well as cancel the resource context.